main.c

```diff
@@ -336,6 +342,18 @@ static void handle_global(void *data, struct wl_registry *registry,
 				&wl_output_interface, 3);
 		surface->output_global_name = name;
 		wl_output_add_listener(surface->output, &_wl_output_listener, surface);
+
+		if (state->zxdg_output_manager) {
+			surface->xdg_output = zxdg_output_manager_v1_get_xdg_output(
+					state->zxdg_output_manager, surface->output);
+			zxdg_output_v1_add_listener(
+					surface->xdg_output, &_xdg_output_listener, surface);
+			wl_display_roundtrip(state->display);
```

Comment:
A roundtrip inside of a Wayland event handle is bad practice and can lead to lockups.

Reply:
Hmm, but the original code does it too? https://github.com/swaywm/swaylock/blob/235b925df7e1bb82d98f1ac8c02e8f85d0a54ee9/main.c#L345

Is it fine when it's protected by the `if (state->run_display)`?

Reply:
The original code does it outside of the event handler.

Reply:
The original code inside the `handle_global` event handler looks like:

``` C
	} else if (strcmp(interface, wl_output_interface.name) == 0) {
		struct swaylock_surface *surface =
			calloc(1, sizeof(struct swaylock_surface));
		surface->state = state;
		surface->output = wl_registry_bind(registry, name,
				&wl_output_interface, 3);
		surface->output_global_name = name;
		wl_output_add_listener(surface->output, &_wl_output_listener, surface);
		wl_list_insert(&state->surfaces, &surface->link);

		if (state->run_display) {
			create_layer_surface(surface);
			wl_display_roundtrip(state->display);
		}
	}
```

Isn't that a `wl_display_roundtrip` in the event handler?

Reply:
Oh well. Then current code is wrong.

In this particular case, I think the existing `wl_display_roundtrip` can be dropped, because there's no code that runs after and depends on some events.

Reply:
Alright. I can try to move the logic out of the event handler.

Reply:
Everything can stay except the roundtrip.